File: Economy/client/items.ts
import {
	type CanOwnMany,
	type CanOwnOne,
	DeserialiseItem,
	IsCanOwnMany,
	IsCanOwnOne,
	IsNumericItem,
	IsStringItem,
	type Item,
	type NumericItem,
	type StringItem,
	TypeNil,
} from "./types"

const errReadOutOfRange = new Error("read out of range")

// fuck Buffer all my homies hate Buffer
export class Buf {
	constructor(public buf: Uint8Array<ArrayBuffer>) { }

	get length() {
		return this.buf.length
	}

	writeUInt8(value: number, offset: number) {
		this.buf[offset] = value
	}

	writeUInt32BE(value: number, offset: number) {
		for (let i = 0; i < 4; i++)
			this.buf[offset + i] = (value >>> (24 - i * 8)) & 0xff
	}

	writeBigUInt64BE(value: bigint, offset: number) {
		const bigValue = BigInt.asUintN(64, value)
		for (let i = 0; i < 8; i++)
			this.buf[offset + i] = Number((bigValue >> BigInt(56 - i * 8)) & 0xffn)
		
	}

	write(str: string, offset: number) {
		const encoder = new TextEncoder()
		const encoded = encoder.encode(str)
		this.buf.set(encoded, offset)
	}

	readUInt8(offset: number): number {
		const v = this.buf[offset]
		if (v === undefined) throw errReadOutOfRange
		return v
	}

	readUInt32BE(offset: number): number {
		let value = 0
		for (let i = 0; i < 4; i++){
			const v = this.buf[offset + i]
			if (v === undefined) throw errReadOutOfRange
			value = (value << 8) | v
		}
		return value
	}

	readBigUInt64BE(offset: number): bigint {
		let value = 0n
		for (let i = 0; i < 8; i++){
			const v = this.buf[offset + i]
			if (v === undefined) throw errReadOutOfRange
			value = (value << 8n) | BigInt(v)
		}
		return value
	}

	subarray(start: number, end?: number): Buf {
		return new Buf(this.buf.subarray(start, end))
	}

	// Hey [object Object] I heard you liked [object Object] so I made you a [object Object] so you can [object Object] while you [object Function] your [object Object]
	toString(): string {
		const decoder = new TextDecoder()
		return decoder.decode(this.buf)
	}

	static concat(buffers: Buf[]): Buf {
		const totalLength = buffers.reduce((sum, b) => sum + b.length, 0)
		const result = new Uint8Array(totalLength)
		let offset = 0
		for (const b of buffers) {
			result.set(b.buf, offset)
			offset += b.length
		}
		return new Buf(result)
	}

	static from(arr: number[] | ArrayBuffer): Buf {
		return new Buf(new Uint8Array(arr))
	}

	static alloc(size: number): Buf {
		return new Buf(new Uint8Array(size))
	}

	static byteLength(str: string): number {
		return new TextEncoder().encode(str).length
	}
}

export function SerialiseUint32(n: number): Buf {
	const buf = Buf.alloc(4)
	buf.writeUInt32BE(n, 0)
	return buf
}

export function SerialiseUint64(n: bigint): Buf {
	const buf = Buf.alloc(8)
	buf.writeBigUInt64BE(n, 0)
	return buf
}

// Serialisation helpers for Item (mirrors ledger/items.go)
function SerialiseNumeric(i: NumericItem): Buf {
	const idbuf = Buf.alloc(5)
	idbuf.writeUInt8(i.Type, 0)
	idbuf.writeUInt32BE(i.ID, 1)
	return idbuf
}

function SerialiseString(i: StringItem): Buf {
	const idstr = i.ID
	const idl = Buf.byteLength(idstr)
	const id = Buf.alloc(2 + idl)
	id.writeUInt8(i.Type, 0)
	id.writeUInt8(idl, 1)
	id.write(idstr, 2)
	return id
}

export function SerialiseItem(i: Item | null): Buf {
	if (i === null) return Buf.from([TypeNil])
	if (IsNumericItem(i)) return SerialiseNumeric(i)
	if (IsStringItem(i)) return SerialiseString(i)
	throw new Error(`unknown Item type: ${i.constructor.name}`)
}

// Simple buffer reader/writer helpers used for serialisation/deserialisation
export class BufReader {
	off = 0
	constructor(private buf: Buf) {}
	read(n: number): Buf {
		if (this.off + n > this.buf.length) throw new Error("read out of range")
		const slice = this.buf.subarray(this.off, this.off + n)
		this.off += n
		return slice
	}
	readUint8(): number {
		const v = this.buf.readUInt8(this.off)
		this.off += 1
		return v
	}
	readUint32(): number {
		const v = this.buf.readUInt32BE(this.off)
		this.off += 4
		return v
	}
	readUint64(): bigint {
		const v = this.buf.readBigUInt64BE(this.off)
		this.off += 8
		return v
	}
	end(): boolean {
		return this.off >= this.buf.length
	}
}

// Items-related types (ported from ledger/items.go)
export type Quantity = bigint

export class ItemsOne {
	set: Set<CanOwnOne>
	constructor(initial?: Iterable<CanOwnOne>) {
		this.set = new Set(initial)
	}

	Empty(): boolean {
		return this.set.size === 0
	}

	[Symbol.iterator]() {
		return this.set[Symbol.iterator]()
	}

	String(): string {
		const parts: string[] = []
		for (const coo of this.set) parts.push(coo.String())
		return `{${parts.join(", ")}}`
	}

	Equal(other: ItemsOne): boolean {
		if (this.set.size !== other.set.size) return false
		for (const coo of this.set) if (!other.set.has(coo)) return false
		return true
	}

	Has(i: CanOwnOne): boolean {
		return this.set.has(i)
	}

	Add(i: CanOwnOne) {
		this.set.add(i)
	}

	Serialise(): Buf {
		const b = []
		const lbuf = Buf.alloc(4)
		lbuf.writeUInt32BE(this.set.size, 0)
		b.push(lbuf)
		for (const i of this.set) {
			b.push(SerialiseItem(i))
		}
		return Buf.concat(b)
	}

	static Deserialise(reader: BufReader): ItemsOne {
		const l = reader.readUint32()
		const io = new ItemsOne()
		for (let idx = 0; idx < l; idx++) {
			const i = DeserialiseItem(reader)
			if (i === null) throw new Error("item is not CanOwnOne: null")
			if (!IsCanOwnOne(i)) throw new Error(`item is not CanOwnOne: ${JSON.stringify(i)}`)
			io.Add(i)
		}
		return io
	}
}

export class ItemsMany {
	map: Map<CanOwnMany, Quantity>
	constructor(initial?: Iterable<[CanOwnMany, Quantity]>) {
		this.map = new Map(initial)
	}

	Empty(): boolean {
		return this.map.size === 0
	}

	[Symbol.iterator]() {
		return this.map[Symbol.iterator]()
	}

	String(): string {
		const parts: string[] = []
		for (const [com, qty] of this.map) parts.push(`${com.String()}: ${qty}`)
		return `{${parts.join(", ")}}`
	}

	Equal(other: ItemsMany): boolean {
		if (this.map.size !== other.map.size) return false
		for (const [com, qty] of this.map) {
			const otherQty = other.map.get(com)
			if (otherQty === undefined || otherQty !== qty) return false
		}
		return true
	}

	Add(i: CanOwnMany, qty: Quantity) {
		// ok maybe I should consider using ?? more. I usually just use || regardless since I don't care about the type conversion, however if one's doing something like `x || ""` it evaluates the rhs even if the lhs is already ""
		const prev = this.map.get(i) ?? 0n
		this.map.set(i, prev + qty)
	}

	Serialise(): Buf {
		const b = []
		const lbuf = Buf.alloc(4)
		lbuf.writeUInt32BE(this.map.size, 0)
		b.push(lbuf)
		for (const [i, qty] of this.map) {
			b.push(SerialiseItem(i))
			const qtybuf = Buf.alloc(8)
			qtybuf.writeBigUInt64BE(qty, 0)
			b.push(qtybuf)
		}
		return Buf.concat(b)
	}

	static Deserialise(reader: BufReader): ItemsMany {
		const l = reader.readUint32()
		const im = new ItemsMany()
		for (let idx = 0; idx < l; idx++) {
			const i = DeserialiseItem(reader)
			if (i === null) throw new Error("item is not CanOwnMany: null")
			if (!IsCanOwnMany(i)) throw new Error(`item is not CanOwnMany: ${JSON.stringify(i)}`)
			const qty = reader.readUint64()
			im.map.set(i, qty)
		}
		return im
	}
}

export class Items {
	One: ItemsOne
	Many: ItemsMany
	constructor(one?: ItemsOne, many?: ItemsMany) {
		this.One = one ?? new ItemsOne()
		this.Many = many ?? new ItemsMany()
	}

	String(): string {
		if (this.One.set.size === 0 && this.Many.map.size === 0)
			return "Items {}"
		let out = "Items {"
		if (this.One.set.size > 0) out += `\n\tOne ${this.One.String()}`
		if (this.Many.map.size > 0) out += `\n\tMany ${this.Many.String()}`
		out += "\n}"
		return out
	}

	Equal(other: Items): boolean {
		return this.One.Equal(other.One) && this.Many.Equal(other.Many)
	}

	IsEmpty(): boolean {
		return this.One.set.size === 0 && this.Many.map.size === 0
	}

	Serialise(): Buf {
		return Buf.concat([this.One.Serialise(), this.Many.Serialise()])
	}

	static Deserialise(reader: BufReader): Items {
		const one = ItemsOne.Deserialise(reader)
		const many = ItemsMany.Deserialise(reader)
		return new Items(one, many)
	}
}

File: Economy/ledger/items.go
package ledger

import (
	"encoding/binary"
	"fmt"
	"io"
	"strings"
)

func SerialiseNumeric(i NumericItem) []byte {
	idbuf := make([]byte, 5)
	idbuf[0] = byte(i.Type())
	binary.BigEndian.PutUint32(idbuf[1:], i.ID())
	return idbuf
}

func SerialiseString(i StringItem) []byte {
	idstr := i.ID()
	idl := len(idstr)
	id := make([]byte, 2+idl)
	id[0] = byte(i.Type())
	id[1] = byte(idl)
	copy(id[2:], idstr)
	return id
}

func SerialiseItem(i Item, w io.Writer) bool {
	if i == nil {
		w.Write([]byte{byte(TypeNil)})
	} else if ni, ok := i.(NumericItem); ok {
		w.Write(SerialiseNumeric(ni))
	} else if si, ok := i.(StringItem); ok {
		w.Write(SerialiseString(si))
	} else {
		return false
	}
	return true
}

func DeserialiseItem(r io.Reader) (Item, error) {
	var typeByte [1]byte
	if _, err := r.Read(typeByte[:]); err != nil {
		return nil, fmt.Errorf("read type: %w", err)
	}

	t := Type(typeByte[0])
	if t == TypeNil {
		return nil, nil
	}

	if t == TypeUser || t == TypeGroup {
		var lenByte [1]byte
		if _, err := r.Read(lenByte[:]); err != nil {
			return nil, fmt.Errorf("read string id length: %w", err)
		}

		idbuf := make([]byte, lenByte[0])
		if _, err := r.Read(idbuf); err != nil {
			return nil, fmt.Errorf("read string id: %w", err)
		}
		id := string(idbuf)

		if t == TypeUser {
			return User{id}, nil
		}
		return Group{id}, nil
	}

	var idbuf [4]byte
	if _, err := r.Read(idbuf[:]); err != nil {
		return nil, fmt.Errorf("read numeric id: %w", err)
	}

	switch id := binary.BigEndian.Uint32(idbuf[:]); t {
	case TypeCurrency:
		return Currency{id}, nil
	case TypeLimitedAsset:
		return LimitedAsset{id}, nil
	case TypeUnlimitedAsset:
		return UnlimitedAsset{id}, nil
	case TypeLimitedSource:
		return LimitedSource{id}, nil
	case TypeUnlimitedSource:
		return UnlimitedSource{id}, nil
	case TypePlace:
		return Place{id}, nil
	}

	return nil, fmt.Errorf("unknown Type: %d", t)
}

type ItemsOne map[CanOwnOne]struct{}

func (io ItemsOne) String() string {
	parts := make([]string, 0, len(io))
	for coo := range io {
		parts = append(parts, coo.String())
	}
	return "{" + strings.Join(parts, ", ") + "}"
}

func (io ItemsOne) Equal(other ItemsOne) bool {
	if len(io) != len(other) {
		return false
	}
	for coo := range io {
		if _, ok := other[coo]; !ok {
			return false
		}
	}
	return true
}

func (io ItemsOne) Has(i CanOwnOne) bool {
	_, ok := io[i]
	return ok
}

func (io *ItemsOne) Add(i CanOwnOne) {
	if *io == nil {
		*io = ItemsOne{i: struct{}{}}
		return
	}
	(*io)[i] = struct{}{}
}

func (io ItemsOne) Serialise(w io.Writer) error {
	var lbuf [4]byte
	// watch out don't have more than 4 billion inventory items lolll
	binary.BigEndian.PutUint32(lbuf[:], uint32(len(io)))
	w.Write(lbuf[:])

	for i := range io {
		if !SerialiseItem(i, w) {
			return fmt.Errorf("unknown CanOwnOne type: %T", i)
		}
	}
	return nil
}

func DeserialiseItemsOne(r io.Reader) (ItemsOne, error) {
	var lbuf [4]byte
	if _, err := r.Read(lbuf[:]); err != nil {
		return nil, fmt.Errorf("read ItemsOne length: %w", err)
	}
	l := binary.BigEndian.Uint32(lbuf[:])

	io := make(ItemsOne, l)
	for range l {
		i, err := DeserialiseItem(r)
		if err != nil {
			return nil, fmt.Errorf("decode item: %w", err)
		}
		coo, ok := i.(CanOwnOne)
		if !ok {
			return nil, fmt.Errorf("item is not CanOwnOne: %T", i)
		}
		io[coo] = struct{}{}
	}
	return io, nil
}

type (
	Quantity  uint64
	ItemsMany map[CanOwnMany]Quantity
)

func (im ItemsMany) String() string {
	parts := make([]string, 0, len(im))
	for com, qty := range im {
		parts = append(parts, fmt.Sprintf("%s: %d", com.String(), qty))
	}
	return "{" + strings.Join(parts, ", ") + "}"
}

func (im ItemsMany) Equal(other ItemsMany) bool {
	if len(im) != len(other) {
		return false
	}
	for com := range im {
		if qty, ok := other[com]; !ok || qty != im[com] {
			return false
		}
	}
	return true
}

func (im *ItemsMany) Add(i CanOwnMany, qty Quantity) {
	if *im == nil {
		*im = ItemsMany{i: qty}
		return
	}
	(*im)[i] += qty
}

func (im ItemsMany) Serialise(w io.Writer) error {
	var lbuf [4]byte
	binary.BigEndian.PutUint32(lbuf[:], uint32(len(im)))
	w.Write(lbuf[:])

	for i, qty := range im {
		if !SerialiseItem(i, w) {
			return fmt.Errorf("unknown CanOwnMany type: %T", i)
		}

		var qtybuf [8]byte
		binary.BigEndian.PutUint64(qtybuf[:], uint64(qty))
		w.Write(qtybuf[:])
	}
	return nil
}

func DeserialiseItemsMany(r io.Reader) (ItemsMany, error) {
	var lbuf [4]byte
	if _, err := r.Read(lbuf[:]); err != nil {
		return nil, fmt.Errorf("read ItemsMany length: %w", err)
	}
	l := binary.BigEndian.Uint32(lbuf[:])

	im := make(ItemsMany, l)
	for range l {
		i, err := DeserialiseItem(r)
		if err != nil {
			return nil, fmt.Errorf("decode item: %w", err)
		}
		com, ok := i.(CanOwnMany)
		if !ok {
			return nil, fmt.Errorf("item is not CanOwnMany: %T", i)
		}

		var qtybuf [8]byte
		if _, err := r.Read(qtybuf[:]); err != nil {
			return nil, fmt.Errorf("read quantity: %w", err)
		}
		qty := binary.BigEndian.Uint64(qtybuf[:])
		im[com] = Quantity(qty)
	}
	return im, nil
}

type Items struct {
	One  ItemsOne
	Many ItemsMany
}

func (is Items) String() string {
	if is.One == nil && is.Many == nil {
		return "Items {}"
	}

	var b strings.Builder
	b.WriteString("Items {")
	if len(is.One) > 0 {
		b.WriteString("\n\tOne ")
		b.WriteString(is.One.String())
	}
	if len(is.Many) > 0 {
		b.WriteString("\n\tMany ")
		b.WriteString(is.Many.String())
	}
	b.WriteString("\n}")
	return b.String()
}

func (is Items) Equal(other Items) bool {
	return is.One.Equal(other.One) && is.Many.Equal(other.Many)
}

func (is Items) IsEmpty() bool {
	return len(is.One) == 0 && len(is.Many) == 0
}

func (is Items) Serialise(w io.Writer) error {
	if err := is.One.Serialise(w); err != nil {
		return fmt.Errorf("serialise ItemsOne: %w", err)
	}

	if err := is.Many.Serialise(w); err != nil {
		return fmt.Errorf("serialise ItemsMany: %w", err)
	}

	return nil
}

func DeserialiseItems(r io.Reader) (is Items, err error) {
	if is.One, err = DeserialiseItemsOne(r); err != nil {
		return Items{}, fmt.Errorf("decode ItemsOne: %w", err)
	}

	if is.Many, err = DeserialiseItemsMany(r); err != nil {
		return Items{}, fmt.Errorf("decode ItemsMany: %w", err)
	}

	return
}
